src/snmp_agent.c

```diff
@@ -0,0 +1,1513 @@
+          return (0);
+        }
+      }
+    }
+  }
+
+  return (0);
+}
+
+static void snmp_agent_free_data(data_definition_t **dd) {
+
+  if (dd == NULL || *dd == NULL)
+    return;
+
+  /* unregister scalar type OID */
+  if ((*dd)->table == NULL) {
+    for (int i = 0; i < (*dd)->oids_len; i++)
+      unregister_mib((*dd)->oids[i].oid, (*dd)->oids[i].oid_len);
+  } else {
+    /* unregister all table OIDs */
+    int *index;
+    char *value;
+
+    c_avl_iterator_t *iter = c_avl_get_iterator((*dd)->table->index_instance);
+    while (c_avl_iterator_next(iter, (void *)&index, (void *)&value) == 0) {
+      for (int i = 0; i < (*dd)->oids_len; i++)
+        snmp_agent_unregister_oid_index(&(*dd)->oids[i], *index);
+    }
+    c_avl_iterator_destroy(iter);
+  }
+
+  sfree((*dd)->name);
+  sfree((*dd)->plugin);
+  sfree((*dd)->plugin_instance);
+  sfree((*dd)->type);
+  sfree((*dd)->type_instance);
+  sfree((*dd)->oids);
+
+  sfree(*dd);
+
+  return;
+}
+
+static void snmp_agent_free_table(table_definition_t **td) {
+
+  if (td == NULL || *td == NULL)
+    return;
+
+  if ((*td)->size_oid.oid_len)
+    unregister_mib((*td)->size_oid.oid, (*td)->size_oid.oid_len);
+
+  if ((*td)->index_oid.oid_len) {
+    int *index;
+    char *value;
+
+    c_avl_iterator_t *iter = c_avl_get_iterator((*td)->index_instance);
+    while (c_avl_iterator_next(iter, (void *)&index, (void *)&value) == 0)
+      snmp_agent_unregister_oid_index(&(*td)->index_oid, *index);
+
+    c_avl_iterator_destroy(iter);
+  }
+
+  for (llentry_t *de = llist_head((*td)->columns); de != NULL; de = de->next) {
+    data_definition_t *dd = (data_definition_t *)de->value;
+    snmp_agent_free_data(&dd);
+  }
+
+  llist_destroy((*td)->columns);
+
+  void *key = NULL;
+  void *value = NULL;
+
+  c_avl_destroy((*td)->index_instance);
+  (*td)->index_instance = NULL;
+
+  while (c_avl_pick((*td)->instance_index, &key, &value) == 0) {
+    sfree(key);
+    sfree(value);
+  }
+  c_avl_destroy((*td)->instance_index);
+  (*td)->instance_index = NULL;
+
+  sfree((*td)->name);
+  sfree(*td);
+
+  return;
+}
+
+static int
+snmp_agent_table_oid_handler(struct netsnmp_mib_handler_s *handler,
+                             struct netsnmp_handler_registration_s *reginfo,
+                             struct netsnmp_agent_request_info_s *reqinfo,
+                             struct netsnmp_request_info_s *requests) {
+
+  if (reqinfo->mode != MODE_GET && reqinfo->mode != MODE_GETNEXT) {
+    DEBUG(PLUGIN_NAME ": Not supported request mode (%d).", reqinfo->mode);
+    return SNMP_ERR_NOERROR;
+  }
+
+  oid_t oid;
+  char temp[DATA_MAX_NAME_LEN];
+
+  pthread_mutex_lock(&g_agent->lock);
+
+  memcpy(oid.oid, requests->requestvb->name,
+         sizeof(oid.oid[0]) * requests->requestvb->name_length);
+  oid.oid_len = requests->requestvb->name_length;
+
+  snmp_agent_oid_to_string(temp, sizeof(temp), &oid);
+
+  DEBUG(PLUGIN_NAME ": Get request received for table OID '%s'.", temp);
```

Comment:
`temp` is not a good name for this variable. How about `oid_str`? It should likely be in an `#if COLLECT_DEBUG` block to avoid unused variable warnings when debugging is disabled.